bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java

```diff
@@ -0,0 +1,285 @@
+        super(30);
+        this.mdnsClient = mdnsClient;
+        this.httpClient = httpClientFactory.getCommonHttpClient();
+        this.descriptionLoader = new DescriptionLoader(httpClient);
+
+        super.activate(configProperties);
+        if (isBackgroundDiscoveryEnabled()) {
+            mdnsClient.addServiceListener(MDNS_SERVICE_TYPE, this);
+        }
+    }
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypes() {
+        return Set.of(THING_TYPE_UID);
+    }
+
+    @Deactivate
+    @Override
+    protected void deactivate() {
+        super.deactivate();
+        mdnsClient.removeServiceListener(MDNS_SERVICE_TYPE, this);
+    }
+
+    @Override
+    public void serviceAdded(@NonNullByDefault({}) ServiceEvent serviceEvent) {
+        considerService(serviceEvent);
+    }
+
+    @Override
+    public void serviceResolved(@NonNullByDefault({}) ServiceEvent serviceEvent) {
+        considerService(serviceEvent);
+    }
+
+    @Override
+    public void serviceRemoved(@NonNullByDefault({}) ServiceEvent serviceEvent) {
+        for (var discoveryResult : discoverWebThing(serviceEvent.getInfo())) {
+            thingRemoved(discoveryResult.getThingUID());
+        }
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        mdnsClient.addServiceListener(MDNS_SERVICE_TYPE, this);
+        startScan(true);
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        mdnsClient.removeServiceListener(MDNS_SERVICE_TYPE, this);
+    }
+
+    private void startScan(boolean isBackground) {
+        scheduler.schedule(() -> scan(isBackground), 0, TimeUnit.SECONDS);
+    }
+
+    @Override
+    protected void startScan() {
+        startScan(false);
+    }
+
+    @Override
+    protected synchronized void stopScan() {
+        removeOlderResults(Instant.now().minus(Duration.ofMinutes(10)).toEpochMilli());
+        super.stopScan();
+    }
+
+    /**
+     * scans the network via mDNS
+     *
+     * @param isBackground true, if is background task
+     */
+    private void scan(boolean isBackground) {
+        var serviceInfos = isBackground ? mdnsClient.list(MDNS_SERVICE_TYPE)
+                : mdnsClient.list(MDNS_SERVICE_TYPE, FOREGROUND_SCAN_TIMEOUT);
+        logger.debug("got {} mDNS entries", serviceInfos.length);
+
+        // create discovery task for each detected service and process these in parallel to increase total discovery
+        // speed
+        var discoveryTasks = Arrays.stream(serviceInfos).map(DiscoveryTask::new).collect(Collectors.toList());
+        try {
+            for (var future : scheduler.invokeAll(discoveryTasks, 5, TimeUnit.MINUTES)) {
+                try {
+                    future.get(5, TimeUnit.MINUTES);
+                } catch (InterruptedException | ExecutionException | TimeoutException e) {
+                    logger.warn("discovering task {} terminated. {}", future, e.getMessage());
+                }
+            }
```

Comment:
If the user cancels the scan process, you need to cancel all tasks.

It is crucial to handle the `InterruptedException` from `loadWebthingDescription()` to finish the tasks immediately when they are cancelled by `stopScan()`.

Reply:
done